text/3-entrees/storage-api/ringbuffer.md

```diff
@@ -0,0 +1,227 @@
+
+_Second Aside: This `B::get()` is one of the reasons for specifying the `Query` associated type on the `StorageValue` type constraint._
+
+### Implementing the `RingBufferTrait`
+We will now implement the `RingBufferTrait`:
+
+```rust, ignore
+impl<Item, B, M, Index> RingBufferTrait<Item> for RingBufferTransient<Item, B, M, Index>
+where // same as the struct definition
+	Item: Codec + EncodeLike,
+	B: StorageValue<(Index, Index), Query = (Index, Index)>,
+	M: StorageMap<Index, Item, Query = Item>,
+	Index: Codec + EncodeLike + Eq + WrappingOps + From<u8> + Copy,
+{
+	fn commit(&self) {
+		B::put((self.start, self.end));
+	}
+```
+
+`commit` just consists of putting the potentially changed bounds into storage. You will notice that we don't update the bounds storage when changing them in the other functions.
+
+```rust, ignore
+	fn is_empty(&self) -> bool {
+		self.start == self.end
+	}
+```
+
+The `is_empty` function just checks whether the start and end bounds have the same value to determine whether the queue is empty, thus avoiding expensive storage accesses. This means we need to uphold the corresponding invariant in the other (notably the `push`) functions.
+
+```rust, ignore
+	fn push(&mut self, item: Item) {
+		M::insert(self.end, item);
+		// this will intentionally overflow and wrap around when bonds_end
+		// reaches `Index::max_value` because we want a ringbuffer.
+		let next_index = self.end.wrapping_add(1.into());
+		if next_index == self.start {
+			// queue presents as empty but is not
+			// --> overwrite the oldest item in the FIFO ringbuffer
+			self.start = self.start.wrapping_add(1.into());
+		}
+		self.end = next_index;
+	}
+```
+
+In the `push` function we insert the pushed `item` into the map and calculate the new bounds by using the `wrapping_add` function. This way our ringbuffer will wrap around when reaching `max_value` of the `Index` type. This is why we need the `WrappingOps` type trait for `Index`.
+
+The `if` is necessary because we need to keep the invariant that `start == end` means that the queue is empty, otherwise we would need to keep track of this state separately. We thus "toss away" the oldest item in the queue if a new item is pushed into a full queue by incrementing the start index.
+
+#### Note: The `WrappingOps` Trait <a name = "wrapping_ops"></a>
+The ringbuffer should be agnostic to the concrete `Index` type used. In order to be able to decrement and increment the start and end index, though, any concrete type needs to implement `wrapping_add` and `wrapping_sub`. Because `std` does not provide such a trait we need another way to require this behavior. One possbility would be using the [`num_traits` crate](https://crates.io/crates/num-traits), but to keep things simple here we just implement our own trait `WrappingOps` for the types we want to support (`u8`, `u16`, `u32` and `u64`).
```

Comment:
Would maybe drop the `####` and make this a `> `
```suggestion
The ringbuffer should be agnostic to the concrete `Index` type used. In order to be able to decrement and increment the start and end index, though, any concrete type needs to implement `wrapping_add` and `wrapping_sub`. Because `std` does not provide such a trait, we need another way to require this behavior. One possbility would be using the [`num_traits` crate](https://crates.io/crates/num-traits), but to keep things simple here we just implement our own trait `WrappingOps` for the types we want to support (`u8`, `u16`, `u32` and `u64`).
```